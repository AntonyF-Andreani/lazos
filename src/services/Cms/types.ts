export interface Component {
  component?: string;
  _uid?: string;
  _editable?: string;
  [key: string]: unknown;
}

export interface ClientLogo extends Component {
  logo: Image;
}

export interface ClientTestimonials extends Component {
  autor: string;
  lugar: string;
  comentario: string;
}

export interface Content {
  banner_image: Image;
  logo_navbar: Image;
  logo_navbar_dark: Image;
  information_section: InformationData[];
  texto_boton?: string;
  sidebar: SidebarData[];
  logo_fundacion: Image;
  logo_andreani: Image;
  title_logos: string;
  terms: File;
  logos: ClientLogo[];
  program_information_title: string;
  program_information_description: string;
  program_information_title_w: string;
  program_information_description_w: string;
  title_steps: string;
  steps: Step[];
  media?: Media[];
  title_testimonials: string;
  testimonials_image: Image;
  testimonials: ClientTestimonials[];
}

export interface Step {
  title: string;
  description: string;
  image: Image;
  second_image?: Image;
}

export interface Media {
  facebook?: string;
  youtube?: string;
  linkedin?: string;
  twitter?: string;
}

export interface Image extends Component {
  filename: string;
  alt?: string;
}

export interface InfoSection {
  link: Link;
}
export interface Link {
  url: string;
}
export interface File extends Component {
  filename: string;
}

export interface InformationData extends InfoSection {
  button_text: string;
  description: string;
}

export interface CmsResponse<T> {
  data: { [key: string]: { content: T; slug?: string; _id?: string; full_slug?: string } };
}

export interface SidebarData {
  _uid?: string;
  link?: Link;
  text: string;
  children?: SidebarData[];
}
