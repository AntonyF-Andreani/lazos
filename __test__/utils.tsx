import React from "react";
import { render } from "@testing-library/react";
import { ThemeProvider } from "@mui/material";

import theme from "theme";

/**
 * Colocar todos los providers, se puede mockear redux y las funciones
 */
const AllTheProviders: React.FC = ({ children }) => {
  return <ThemeProvider theme={theme}>{children}</ThemeProvider>;
};

const customRender = (ui: React.ReactElement) => render(ui, { wrapper: AllTheProviders });

// re-export everything
export * from "@testing-library/react";

// override render method
export { customRender as render };
