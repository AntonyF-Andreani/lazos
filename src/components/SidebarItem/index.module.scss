.listItem {
  &:hover {
    color: var(--primary);
    cursor: pointer;
  }
}

.listItemText {
  font-style: normal;
  line-height: 20px;
  font-family: "Work Sans", Arial, sans-serif;
  text-transform: uppercase;
  font-size: 12px;
  font-weight: 700;
  text-align: left;
  width: 100%;
}

.sidebarItemChild {
  font-size: 0.8em;
  font-weight: 400;
  letter-spacing: -0.5px;
  padding-block: 5px;
  color: gray;

  &:hover {
    background-color: rgb(0 0 0 / 0.05);
  }

  & > div > span {
    text-transform: capitalize;
    font-weight: 400;
  }
}

.listItemIcon {
  margin-left: auto;
  transition: margin-right 0.3s ease-in-out;
  color: inherit;

  & > svg {
    margin-right: 5px;
    margin-left: auto;
    transition: transform 0.3s ease-in-out;
  }

  &[data-toright="true"] {
    margin-right: -8px;
  }
}

.list {
  margin-left: 10px;
}