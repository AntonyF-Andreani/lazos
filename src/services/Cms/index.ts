import axios, { AxiosInstance, AxiosResponse } from "axios";

import { Content, CmsResponse } from "./types";

class CmsServiceClass {
  private readonly client: AxiosInstance;

  constructor() {
    this.client = axios.create({
      baseURL: "https://gapi.storyblok.com/v1/api",
      headers: {
        token: process.env.NEXT_PUBLIC_STORYBLOK_TOKEN as string,
        "Content-Type": "application/json",
        Accept: "application/json",
      },
    });

    this.client.interceptors.response.use((resp: AxiosResponse<{ data: { error?: string } }>) => {
      if (resp.data.data.error) {
        throw new Error(resp.data.data.error);
      }

      return resp;
    });
  }

  getContentFetch = async () =>
    this.client.post<{ query: string }, AxiosResponse<CmsResponse<Content>>>("", {
      query: `{
          LazoscontentItem(id: "lazos/instituciones"){
            content {
              banner_image{
                alt
                filename
              }
              information_section
              logo_navbar {
                alt
                filename
                title
              }
              logo_navbar_dark {
                alt
                filename
                title
              }
              sidebar
              logo_fundacion{
                alt
                filename
              }
              logo_andreani{
                alt
                filename
              }
              title_logos
              terms {
                filename
              }
              logos
              program_information_title
              program_information_description              
              program_information_title_w
              program_information_description_w
              media
              title_steps,
              steps
              title_testimonials
              testimonials_image {
                alt
                filename                
              }
              testimonials
            }
          }
        }`,
    });
}

export const CmsService = new CmsServiceClass();
