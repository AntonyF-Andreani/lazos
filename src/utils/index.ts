import { ImageLoaderProps } from "next/image";

/**
 * Utilidad para extraerla URL con formatos optimizados para las imagenes del CMS (Servicio de Imagenes)
 * https://www.storyblok.com/docs/image-service
 */
export const StoryBlockImageURL = ({
  src,
  width,
  quality = 20,
  format = "webp",
  height = 0,
}: StoryBlockImageURLParams) => {
  return `${src}/m/${width}x${height}/filters:quality(${quality}):format(${format})`;
};

interface StoryBlockImageURLParams {
  /** URL de la imagen como la provee Storyblok */
  src: string;
  /** Width of image. The Storyblok API infers aspect ratio */
  width: number;
  /** Quality of image, for default its 20 (20%). */
  quality?: number;
  /** Format of image, for default its webp */
  format?: "webp" | "jpeg" | "png";
  /** Height of the image */
  height?: number;
}

export const StoryBlockLoader = ({ src, width, quality = 20 }: ImageLoaderProps) => {
  return StoryBlockImageURL({ src, width, quality });
};

/**
 * Util for join classes undefined or string and filter undefined
 * @param args classNames
 *
 *  @example const className = joinClasses("a", "b", undefined, "c") // "a b c"
 */
export const joinClasses = (...args: (string | undefined)[]): string => {
  return args.filter((x) => x).join(" "); // undefined is falsey
};

/**
 * Util to test diferrent viewports, must be used with jest.
 *
 * This util make a mock implementation of window.matchMedia
 *
 * @example prepareViewport("(max-width: 768px)")
 */
export const prepareViewport = (mediaQuery: string) => {
  Object.defineProperty(window, "watchMedia", {
    writable: true,
    configurable: true,
    value: mediaQuery,
  });

  window.matchMedia = jest.fn().mockImplementation((query) => {
    return {
      matches: query === mediaQuery,
      media: "",
      onchange: null,
      addListener: jest.fn(),
      removeListener: jest.fn(),
    };
  });
};
