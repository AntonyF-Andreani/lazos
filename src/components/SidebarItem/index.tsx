import { faChevronRight } from "@fortawesome/pro-light-svg-icons";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import {
  Collapse,
  Link,
  List,
  ListItem,
  ListItemIcon,
  ListItemIconProps,
  ListItemProps,
  ListItemText,
  ListItemTextProps,
  ListProps,
} from "@mui/material";
import React from "react";

import styles from "./index.module.scss";

import { joinClasses } from "utils";
import { SidebarData } from "services/Cms/types";

export const SidebarItemAdapter = (sidebarData: SidebarData): SidebarItemProps => {
  return {
    listItemTextProps: {
      primary: sidebarData.text,
    },
    url: sidebarData.link?.url,
    subItems: sidebarData.children?.map(SidebarItemAdapter),
  };
};

export interface SidebarItemProps extends ListItemProps {
  listItemTextProps?: ListItemTextProps;
  listItemIconStartProps?: ListItemIconProps;
  subItems?: SidebarItemProps[];
  listProps?: ListProps;
  url?: string;
}

const SidebarItem = ({
  listItemTextProps,
  listItemIconStartProps,
  subItems,
  listProps,
  url,
  ...props
}: SidebarItemProps) => {
  const [isOpen, setIsOpen] = React.useState(false);
  const [isEnter, setIsEnter] = React.useState(false);

  const handleToggle = () => {
    setIsOpen((prev) => !prev);
  };

  return (
    <>
      <Link
        color="inherit"
        data-testid="sidebar-item-url"
        href={url !== "" ? url : undefined}
        underline="none"
      >
        <ListItem
          className={joinClasses(styles.listItem, props.className)}
          onClick={subItems ? handleToggle : undefined}
          onMouseEnter={() => setIsEnter(true)}
          onMouseLeave={() => setIsEnter(false)}
          {...props}
        >
          {listItemIconStartProps && <ListItemIcon {...listItemIconStartProps} />}
          {listItemTextProps && (
            <ListItemText
              color="inherit"
              {...listItemTextProps}
              data-testid="sidebar-item-text"
              primaryTypographyProps={{
                className: joinClasses(
                  styles.listItemText,
                  listItemTextProps.primaryTypographyProps?.className
                ),
              }}
            />
          )}
          {subItems && subItems.length > 0 && (
            <ListItemIcon
              className={styles.listItemIcon}
              data-testid="sidebar-item-icon"
              data-toright={isEnter || isOpen}
              sx={{
                marginRight: isEnter || isOpen ? "-8px" : 0,
              }}
            >
              <FontAwesomeIcon
                color="inherit"
                icon={faChevronRight}
                style={{
                  transform: isOpen ? "rotate(90deg)" : "rotate(0deg)",
                }}
              />
            </ListItemIcon>
          )}
        </ListItem>
      </Link>
      {subItems && subItems.length > 0 && (
        <Collapse in={isOpen}>
          <List className={styles.list} data-testid="sidebar-subitems-list" {...listProps}>
            {subItems.map((subItem, i) => (
              <SidebarItem
                key={`SidebarSubItem-${i}-${subItem.listItemTextProps?.primary ?? ""}`}
                className={joinClasses(styles.sidebarItemChild, subItem.className)}
                data-testid={`SidebarSubItem-${i}-${subItem.listItemTextProps?.primary ?? ""}`}
                {...subItem}
              />
            ))}
          </List>
        </Collapse>
      )}
    </>
  );
};

export default SidebarItem;
