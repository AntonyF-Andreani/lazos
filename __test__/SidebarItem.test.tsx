import React from "react";
import { fireEvent } from "@testing-library/react";

import { render } from "./utils";

import SidebarItem, { SidebarItemAdapter } from "components/SidebarItem";
import { SidebarData } from "services/Cms/types";

describe("<SidebarItem />", () => {
  it("should render", () => {
    const { asFragment } = render(<SidebarItem />);

    expect(asFragment()).toMatchSnapshot();
  });
  it("should render children", () => {
    const { getByTestId, getByText } = render(
      <SidebarItem subItems={[{ listItemTextProps: { primary: "test" } }]} />
    );

    const text = getByText("test");
    const subItemsIcon = getByTestId("sidebar-item-icon");
    const listSubItems = getByTestId("sidebar-subitems-list");

    expect(text).toBeInTheDocument();
    expect(subItemsIcon).toBeInTheDocument();
    expect(listSubItems).toBeInTheDocument();
  });
  it("should render with classNames", () => {
    const { getByTestId } = render(
      <SidebarItem
        listItemTextProps={{
          primary: "test",
          primaryTypographyProps: { className: "testClass" },
        }}
      />
    );

    const text = getByTestId("sidebar-item-text");

    expect(text.firstChild).toHaveClass("testClass");
    expect(text.firstChild).toHaveClass("listItemText");
  });
  it("should render children with idGenerated", () => {
    const { getByTestId } = render(
      <SidebarItem subItems={[{ listItemTextProps: { primary: "test" } }]} />
    );
    const testIdGenerated = getByTestId("SidebarSubItem-0-test");

    expect(testIdGenerated).toBeInTheDocument();
  });
  it("should render children with idGenerated and without text", () => {
    const { getByTestId } = render(<SidebarItem subItems={[{}]} />);
    const testIdGenerated = getByTestId("SidebarSubItem-0-");

    expect(testIdGenerated).toBeInTheDocument();
  });
  it("should not render arrow and list with empty data", () => {
    const wrapper = render(<SidebarItem subItems={[]} />);

    const { queryAllByTestId } = wrapper;
    const subItemsIcon = queryAllByTestId("sidebar-item-icon");
    const listSubItems = queryAllByTestId("sidebar-subitems-list");

    expect(subItemsIcon).toHaveLength(0);
    expect(listSubItems).toHaveLength(0);
  });
  it("should not render arrow and list with undefined subItems", () => {
    const wrapper = render(<SidebarItem />);

    const { queryAllByTestId } = wrapper;
    const subItemsIcon = queryAllByTestId("sidebar-item-icon");
    const listSubItems = queryAllByTestId("sidebar-subitems-list");

    expect(subItemsIcon).toHaveLength(0);
    expect(listSubItems).toHaveLength(0);
  });

  it("Should url be undefined if is not in props", () => {
    const { getByTestId } = render(<SidebarItem />);

    const element = getByTestId("sidebar-item-url");

    expect(element.getAttribute("href")).toBeNull();
  });

  it("Should url be undefined if is an empty string", () => {
    const { getByTestId } = render(<SidebarItem url="" />);

    const element = getByTestId("sidebar-item-url");

    expect(element.getAttribute("href")).toBeNull();
  });

  it("Should url be url", () => {
    const url = "url";
    const { getByTestId } = render(<SidebarItem url={url} />);

    const element = getByTestId("sidebar-item-url");

    expect(element.getAttribute("href")).toBe(url);
  });

  it("Should have icon when is passed by prop", () => {
    const url = "url";
    const { getByTestId } = render(
      <SidebarItem
        listItemIconStartProps={{
          children: <div data-testid="icon-test" />,
        }}
        url={url}
      />
    );

    const startIcon = getByTestId("icon-test");

    expect(startIcon).toBeInTheDocument();
  });
  it("Should have styles  on open", () => {
    const { getByTestId } = render(
      <SidebarItem subItems={[{ listItemTextProps: { primary: "test" } }]} />
    );
    const sidebarItemIcon = getByTestId("sidebar-item-icon");

    fireEvent.click(sidebarItemIcon);

    expect(sidebarItemIcon.firstChild).toHaveStyle("transform: rotate(90deg)");
    expect(sidebarItemIcon).toHaveStyle("margin-right: -8px");
  });
  it("Should have styles  on hover", () => {
    const { getByTestId } = render(
      <SidebarItem
        data-testid="sidebarItem"
        subItems={[{ listItemTextProps: { primary: "test" } }]}
      />
    );
    const sidebarItemIcon = getByTestId("sidebar-item-icon");
    const sidebarItem = getByTestId("sidebarItem");

    fireEvent.mouseEnter(sidebarItem);

    expect(sidebarItemIcon).toHaveStyle("margin-right: -8px");
    fireEvent.mouseLeave(sidebarItem);

    expect(sidebarItemIcon).toHaveStyle("margin-right: 0px");
  });
});

describe("SidebarItemAdapter()", () => {
  it("Shold return basic object", () => {
    const sidebarData: SidebarData = {
      text: "text",
      _uid: "uid",
    };

    const result = SidebarItemAdapter(sidebarData);

    expect(result).toEqual({
      listItemTextProps: {
        primary: "text",
      },
    });
  });
  it("Should return url property when exists", () => {
    const sidebarData: SidebarData = {
      text: "text",
      _uid: "uid",
      link: {
        url: "url",
      },
    };

    const result = SidebarItemAdapter(sidebarData);

    expect(result).toEqual({
      listItemTextProps: {
        primary: "text",
      },
      url: "url",
    });
  });
  it("Should return children", () => {
    const sidebarData: SidebarData = {
      text: "text",
      _uid: "uid",
      children: [
        {
          _uid: "casea",
          text: "child1",
        },
      ],
    };

    const result = SidebarItemAdapter(sidebarData);

    expect(result).toEqual({
      listItemTextProps: {
        primary: "text",
      },
      subItems: [
        {
          listItemTextProps: {
            primary: "child1",
          },
        },
      ],
    });
  });
});
