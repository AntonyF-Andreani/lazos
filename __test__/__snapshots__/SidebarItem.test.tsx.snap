// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SidebarItem /> should render 1`] = `
<DocumentFragment>
  <a
    class="MuiTypography-root MuiTypography-inherit MuiLink-root MuiLink-underlineNone css-1n8yj6x-MuiTypography-root-MuiLink-root"
    data-testid="sidebar-item-url"
  >
    <li
      class="MuiListItem-root MuiListItem-gutters MuiListItem-padding listItem css-mxv0kx-MuiListItem-root"
    />
  </a>
</DocumentFragment>
`;
